{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "importing Jupyter notebook from data_transformation.ipynb\n"
     ]
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "import sys\n",
    "import argparse\n",
    "\n",
    "import numpy\n",
    "import pandas\n",
    "\n",
    "import mhcnames\n",
    "\n",
    "import Bio.SeqIO  # pylint: disable=import-error\n",
    "import import_ipynb\n",
    "import data_transformation\n",
    "\n",
    "def normalize_simple(s):\n",
    "    return mhcnames.normalize_allele_name(s)\n",
    "\n",
    "\n",
    "def normalize_complex(s, disallowed=[\"MIC\", \"HFE\"]):\n",
    "    if any(item in s for item in disallowed):\n",
    "        return None\n",
    "    try:\n",
    "        return normalize_simple(s)\n",
    "    except:\n",
    "        while s:\n",
    "            s = \":\".join(s.split(\":\")[:-1])\n",
    "            try:\n",
    "                return normalize_simple(s)\n",
    "            except:\n",
    "                pass\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_to_sequence = {}\n",
    "reader = Bio.SeqIO.parse(args.aligned_fasta, \"fasta\")\n",
    "for record in reader:\n",
    "    name = record.description.split()[1]\n",
    "    print(record.name, record.description)\n",
    "    allele_to_sequence[name] = str(record.seq)\n",
    "\n",
    "print(\"Read %d aligned sequences\" % len(allele_to_sequence))\n",
    "\n",
    "allele_sequences = pandas.Series(allele_to_sequence).to_frame()\n",
    "allele_sequences.columns = ['aligned']\n",
    "allele_sequences['aligned'] = allele_sequences['aligned'].str.replace(\n",
    "    \"-\", \"X\")\n",
    "\n",
    "allele_sequences['normalized_allele'] = allele_sequences.index.map(normalize_complex)\n",
    "allele_sequences = allele_sequences.set_index(\"normalized_allele\", drop=True)\n",
    "\n",
    "selected_positions = []\n",
    "\n",
    "recapitulate_df = pandas.read_csv(args.recapitulate_sequences)\n",
    "recapitulate_df[\"normalized_allele\"] = recapitulate_df.allele.map(\n",
    "    normalize_complex)\n",
    "recapitulate_df = (\n",
    "    recapitulate_df\n",
    "        .dropna()\n",
    "        .drop_duplicates(\"normalized_allele\")\n",
    "        .set_index(\"normalized_allele\", drop=True))\n",
    "\n",
    "allele_sequences[\"recapitulate_target\"] = recapitulate_df.iloc[:,-1]\n",
    "\n",
    "print(\"Sequences in recapitulate CSV that are not in aligned fasta:\")\n",
    "print(recapitulate_df.index[\n",
    "    ~recapitulate_df.index.isin(allele_sequences.index)\n",
    "].tolist())\n",
    "\n",
    "allele_sequences_with_target = allele_sequences.loc[\n",
    "    ~allele_sequences.recapitulate_target.isnull()\n",
    "]\n",
    "\n",
    "position_identities = []\n",
    "target_length = int(\n",
    "    allele_sequences_with_target.recapitulate_target.str.len().max())\n",
    "for i in range(target_length):\n",
    "    series_i = allele_sequences_with_target.recapitulate_target.str.get(i)\n",
    "    row = []\n",
    "    full_length_sequence_length = int(\n",
    "        allele_sequences_with_target.aligned.str.len().max())\n",
    "    for k in range(full_length_sequence_length):\n",
    "        series_k = allele_sequences_with_target.aligned.str.get(k)\n",
    "        row.append((series_i == series_k).mean())\n",
    "    position_identities.append(row)\n",
    "\n",
    "position_identities = pandas.DataFrame(numpy.array(position_identities))\n",
    "selected_positions = position_identities.idxmax(1).tolist()\n",
    "fractions = position_identities.max(1)\n",
    "print(\"Selected positions: \", *selected_positions)\n",
    "print(\"Lowest concordance fraction: %0.5f\" % fractions.min())\n",
    "assert fractions.min() > 0.99\n",
    "\n",
    "allele_sequences[\"recapitulated\"] = allele_sequences.aligned.map(\n",
    "    lambda s: \"\".join(s[p] for p in selected_positions))\n",
    "\n",
    "allele_sequences_with_target = allele_sequences.loc[\n",
    "    ~allele_sequences.recapitulate_target.isnull()\n",
    "]\n",
    "\n",
    "agreement = (\n",
    "    allele_sequences_with_target.recapitulated ==\n",
    "    allele_sequences_with_target.recapitulate_target).mean()\n",
    "\n",
    "print(\"Overall agreement: %0.5f\" % agreement)\n",
    "assert agreement > 0.9\n",
    "\n",
    "# Add additional positions\n",
    "if args.differentiate_alleles:\n",
    "    differentiate_alleles = pandas.read_csv(\n",
    "        args.differentiate_alleles).iloc[:,0].values\n",
    "    print(\n",
    "        \"Read %d alleles to differentiate:\" % len(differentiate_alleles),\n",
    "        differentiate_alleles)\n",
    "\n",
    "    allele_sequences_to_differentiate = allele_sequences.loc[\n",
    "        allele_sequences.index.isin(differentiate_alleles)\n",
    "    ]\n",
    "    print(allele_sequences_to_differentiate.shape)\n",
    "\n",
    "    additional_positions = []\n",
    "    for (_, sub_df) in allele_sequences_to_differentiate.groupby(\"recapitulated\"):\n",
    "        if sub_df.aligned.nunique() > 1:\n",
    "            differing = pandas.DataFrame(\n",
    "                dict([(pos, chars) for (pos, chars) in\n",
    "                enumerate(zip(*sub_df.aligned.values)) if\n",
    "                any(c != chars[0] for c in chars) and \"X\" not in chars])).T\n",
    "            print(sub_df)\n",
    "            print(differing)\n",
    "            print()\n",
    "            additional_positions.extend(differing.index)\n",
    "\n",
    "additional_positions = sorted(set(additional_positions))\n",
    "print(\n",
    "    \"Selected %d additional positions: \" % len(additional_positions),\n",
    "    additional_positions)\n",
    "\n",
    "extended_selected_positions = sorted(\n",
    "    set(selected_positions).union(set(additional_positions)))\n",
    "print(\n",
    "    \"Extended selected positions (%d)\" % len(extended_selected_positions),\n",
    "    *extended_selected_positions)\n",
    "\n",
    "allele_sequences[\"sequence\"] = allele_sequences.aligned.map(\n",
    "    lambda s: \"\".join(s[p] for p in extended_selected_positions))\n",
    "\n",
    "allele_sequences[[\"sequence\"]].to_csv(args.out_csv, index=True)\n",
    "print(\"Wrote: %s\" % args.out_csv)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
